src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -129,29 +130,61 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                         throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
                     }
 
-                    if (name.SequenceEqual(s_statusHeaderName))
+                    if (name[0] == 58) // ':'
                     {
                         if (_state == StreamState.ExpectingTrailingHeaders)
                         {
                             // Pseudo-headers not allowed in trailers.
                             if (NetEventSource.IsEnabled) _connection.Trace("Pseudo-header in trailer headers.");
                             throw new HttpRequestException(SR.net_http_invalid_response);
                         }
-
-                        if (value.Length != 3)
-                            throw new Exception("Invalid status code");
-
-                        // Copied from HttpConnection
-                        byte status1 = value[0], status2 = value[1], status3 = value[2];
-                        if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                        if (name.SequenceEqual(s_statusHeaderName))
+                        {
+                            if (_response.StatusCode != 0)
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace("Received duplicate status headers.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            if (value.Length != 3)
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace($"Received status headers '{System.Text.Encoding.ASCII.GetString(value)}'.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            // Copied from HttpConnection
+                            byte status1 = value[0];
+                            byte status2 = value[1];
+                            byte status3 = value[2];
+                            if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace($"Invalid status header '{System.Text.Encoding.ASCII.GetString(value)}'.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            int statusValue = (100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0'));
+                            if (statusValue == 0)
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace($"Invalid status header '{System.Text.Encoding.ASCII.GetString(value)}'.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
```

Comment:
This seems wrong to me.  We don't fail for a 000 response status code with HTTP/1.1.  Is there a reason we would for 2.0?
```C#
using System;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Threading.Tasks;

class Program
{
    static async Task Main()
    {
        using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
        {
            listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
            listener.Listen(1);

            var hc = new HttpClient();

            Task<Socket> serverTask = listener.AcceptAsync();
            Task<HttpResponseMessage> responseTask = hc.GetAsync($"http://localhost:{((IPEndPoint)listener.LocalEndPoint).Port}");

            using (Socket server = await serverTask)
            using (StreamReader reader = new StreamReader(new NetworkStream(server)))
            using (StreamWriter writer = new StreamWriter(new NetworkStream(server)) { AutoFlush = true })
            {
                while (!string.IsNullOrEmpty(reader.ReadLine())) ;
                writer.Write($"HTTP/1.1 000 OK\r\nContent-Length: 5\r\n\r\nHello");
                server.Shutdown(SocketShutdown.Both);

                using (HttpResponseMessage m = await responseTask)
                {
                    Console.WriteLine(m.StatusCode);
                    Console.WriteLine(m.ReasonPhrase);
                }
            }
        }
    }
}
```
That runs successfully (outputting `0` and `OK`) on both .NET Core and .NET Framework.

Reply:
Status codes < 100 are not defined by the RFC:
https://tools.ietf.org/html/rfc7231#section-6

And this text doesn't help with dealing with status codes < 100 since it doesn't have the format of x00:

```
   HTTP status codes are extensible.  HTTP clients are not required to
   understand the meaning of all registered status codes, though such
   understanding is obviously desirable.  However, a client MUST
   understand the class of any status code, as indicated by the first
   digit, and treat an unrecognized status code as being equivalent to
   the x00 status code of that class, with the exception that a
   recipient MUST NOT cache a response with an unrecognized status code.

   For example, if an unrecognized status code of 471 is received by a
   client, the client can assume that there was something wrong with its
   request and treat the response as if it had received a 400 (Bad
   Request) status code.  The response message will usually contain a
   representation that explains the status.
```

We should probably do what .NET Framework is currently doing unless we think we are adding value by throwing an exception here.

Reply:
I could add bool or add new State to track this. I was cheap and decided to overload status code to see if we received valid :status.

Alternatively, we can delay `_response = new HttpResponseMessage()` from constructor as :status must be first header in response.